File: Lab3Java/src/TVector3D.java
import java.util.Objects;

public class TVector3D extends TVector2D {
    static final String[] coordinateNotations = {"x", "y", "z"};

    private double z;

    TVector3D() {
        this(0, 0, 0);
    }

    TVector3D(double[] coordinates) {
        super(coordinates[0], coordinates[1]);
        z = coordinates[2];
    }

    TVector3D(double x, double y, double z) {
        super(x, y);
        this.z = z;
    }

    double getZ() {
        return z;
    }

    void setZ(double z) {
        this.z = z;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof TVector3D)) return false;
        if (!super.equals(o)) return false;
        TVector3D tVector3D = (TVector3D) o;
        return Double.compare(tVector3D.z, z) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), z);
    }

    @Override
    public String toString() {
        return "TVector3D{" +
                "x=" + getX() +
                ", y=" + getY() +
                ", z=" + z +
                '}';
    }

    @Override
    double getLength() {
        return super.getLength() + Math.sqrt(Math.pow(z, 2));
    }

    @Override
    TVector3D normalize() {
        return new TVector3D(getX() / getLength(), getY() / getLength(), z / getLength());
    }

    TVector3D plus(TVector3D v2) {
        return new TVector3D(super.getX() + v2.getX(), super.getY() + v2.getY(), z + getZ());
    }

    TVector3D minus(TVector3D v2) {
        return new TVector3D(super.getX() - v2.getX(), super.getY() - v2.getY(), z - v2.getZ());
    }

    TVector3D multiply(double s2) {
        return new TVector3D(getX() * s2, getY() * s2, z * s2);
    }

    TVector3D divide(double s2) {
        return new TVector3D(getX() / s2, getY() / s2, z / s2);
    }
}

File: Lab3Csharp/Lab3Csharp/TVector3D.cs
using System;

namespace Lab3Csharp
{
    internal class TVector3D : TVector2D
    {
        public new static readonly string[] coordinateNotations = { "x", "y", "z" };

        public double Z { get; private set; }

        public TVector3D() : this(0, 0, 0) { }

        public TVector3D(double[] cordinates) : this(cordinates[0], cordinates[1], cordinates[2]) { }

        public TVector3D(double x, double y, double z) { X = x; Y = y; Z = z; }

        public TVector3D(TVector3D v1) { X = v1.X; Y = v1.Y; Z = v1.Z; }

        public override string ToString() => base.ToString() +"\tz: " + Z;

        public override bool Equals(object obj) => obj == null ? false : Equals((TVector3D)obj);

        public bool Equals(TVector3D vector) => X == vector.X && Y == vector.Y && Z == vector.Z;

        public override int GetHashCode()
        {
            int hashCode = X.GetHashCode();
            hashCode = 31 * hashCode + Y.GetHashCode();
            hashCode = 31 * hashCode + Z.GetHashCode();
            return hashCode;
        }

        public override double GetLength() => Math.Sqrt(Math.Pow(X, 2) + Math.Pow(Y, 2) + Math.Pow(Z, 2));

        public override TVector2D Normalize() => new TVector3D(X / GetLength(), Y / GetLength(), Z / GetLength());

        public static bool operator ==(TVector3D v1, TVector3D v2) => v1.X == v2.X && v1.Y == v2.Y && v1.Z == v2.Z;

        public static bool operator !=(TVector3D v1, TVector3D v2) => !(v1 == v2);

        public static TVector3D operator +(TVector3D v1, TVector3D v2) => new TVector3D(v1.X + v2.X, v1.Y + v2.Y, v1.Z + v2.Z);

        public static TVector3D operator -(TVector3D v1, TVector3D v2) => new TVector3D(v1.X - v2.X, v1.Y - v2.Y, v1.Z - v2.Z);

        public static TVector3D operator *(TVector3D v1, double s2) => new TVector3D(v1.X * s2, v1.Y * s2, v1.Z * s2);

        public static TVector3D operator /(TVector3D v1, double s2) => new TVector3D(v1.X / s2, v1.Y / s2, v1.Z / s2);
    }
}